File: src/test/smallTestProject/AlarmClock/Foundation/src/main/java/com/h2m/common/observer/Observable.java
package com.h2m.common.observer;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class Observable
{
    public interface IObserver
    {
        public void handleEvent(Observable observable, String event);
    }

    private final Map<String, List<IObserver>> m_EventToObservers = new HashMap<String, List<IObserver>>();

    static
    {
    }
    
    protected Observable(String... events)
    {
        assert events != null : "'events' must not be null";
        assert events.length > 0 : "At least one event needs to be defined";
        for (String nextEvent : events)
        {
            List<IObserver> previous = m_EventToObservers.put(nextEvent, new ArrayList<IObserver>());
            assert previous == null : "Duplicate event defined: " + nextEvent;
        }
    }

    public final void addObserver(IObserver observer, String event)
    {
        assert observer != null : "'observer' must not be null";
        assert event != null : "'event' must not be null";
        assert m_EventToObservers.containsKey(event) : "Event not supported: " + event;
        assert !m_EventToObservers.get(event).contains(observer) : "Observer already added" + observer;
        m_EventToObservers.get(event).add(observer);
    }

    protected final void notifyAboutEvent(String event)
    {
        assert event != null : "'event' must not be null";
        assert m_EventToObservers.containsKey(event) : "Event not supported: " + event;
        for (IObserver nextObserver : m_EventToObservers.get(event))
        {
            nextObserver.handleEvent(this, event);
        }
    }
}

File: src/test/NAlarmClock/Foundation/Observable.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;

namespace Foundation
{
	public abstract class Observable
	{
		public interface IObserver
		{
			void handleEvent(Observable observable, String eventString);
		}

		private Dictionary<String, List<IObserver>> m_EventToObservers = new Dictionary<String, List<IObserver>>();

		protected Observable(params String[] events)
		{
			Debug.Assert(events != null, "'events' must not be null");
			Debug.Assert(events.Length > 0, "At least one event needs to be defined");
			foreach (String nextEvent in events)
			{
				Debug.Assert(m_EventToObservers.ContainsKey(nextEvent), "Duplicate event defined: " + nextEvent);
				m_EventToObservers.Add(nextEvent, new List<IObserver>());
			}
		}

		public virtual void addObserver(IObserver observer, String eventString)
		{
			Debug.Assert(observer != null, "'observer' must not be null");
			Debug.Assert(eventString != null, "'event' must not be null");
			Debug.Assert(m_EventToObservers.ContainsKey(eventString), "Event not supported: " + eventString);
			List<IObserver> list = m_EventToObservers[eventString];
			Debug.Assert(!list.Contains(observer),"Observer already added" + observer);
			list.Add (observer);
		}

		protected virtual void notifyAboutEvent(String eventString)
		{
			Debug.Assert(eventString != null, "'event' must not be null");
			Debug.Assert(m_EventToObservers.ContainsKey(eventString), "Event not supported: " + eventString);
			foreach (IObserver nextObserver in m_EventToObservers[eventString])
			{
				nextObserver.handleEvent(this, eventString);
			}
		}
	}
}




